source/text/Format.cpp

```diff
@@ -318,38 +406,30 @@ string Format::Replace(const string &source, const map<string, string> &keys)
 	string result;
 	result.reserve(source.length());
 
-	size_t start = 0;
-	size_t search = start;
-	while(search < source.length())
+	auto ProcessText = [&](const string &from, size_t start, size_t length)
 	{
-		size_t left = source.find('<', search);
-		if(left == string::npos)
-			break;
+		result.append(from, start, length);
+	};
+	auto SubstitutionFor = [&](const string &key) -> const string *
+	{
+		auto found = keys.find(key);
+		return (found == keys.end()) ? nullptr : &found->second;
+	};
 
-		size_t right = source.find('>', left);
-		if(right == string::npos)
-			break;
+	ScanForSubstitutions(source, ProcessText, SubstitutionFor);
+	return result;
+}
 
-		bool matched = false;
-		++right;
-		size_t length = right - left;
-		for(const auto &it : keys)
-			if(!source.compare(left, length, it.first))
-			{
-				result.append(source, start, left - start);
-				result.append(it.second);
-				start = right;
-				search = start;
-				matched = true;
-				break;
-			}
 
-		if(!matched)
-			search = left + 1;
-	}
 
-	result.append(source, start, source.length() - start);
-	return result;
+void Format::Expand(map<string, string> &keys)
+{
+	map<string, string> newKeys;
+	unordered_set<string> keysBeingExpanded;
+	for(auto it = keys.begin(); it != keys.end(); ++it)
+		if(newKeys.find(it->first) == newKeys.end())
```

Comment:
```suggestion
		if(!newKeys.count(it->first))
```

Reply:
No. Count counts the number of keys. For a map, this is a trivial operation. For anything else, it'll become O(N). Using count in place of find is a bad practice.